solana/pyth2wormhole/client/src/main.rs

```diff
@@ -190,301 +225,276 @@ fn handle_attest(
+            "{} of {} batches failed:\n{}",
+            errors.len(),
+            batch_count,
+            err_lines
+        );
+        error!("{}", msg);
+        return Err(msg.into());
+    }
 
-        // Print stats on every pass through the batches
-        let stats_seconds_elapsed = stats_start_time.elapsed().as_millis() as f64 / 1000.0;
-        let tps = tx_successes as f64 / stats_seconds_elapsed;
-        let sym_ps = tps * config.max_batch_size as f64;
+    Ok(())
+}
 
-        let mut total = (tx_successes + tx_send_failures + tx_confirm_failures) as f64;
-        // Avoid division by 0
-        if total < 0.0001 {
-            total = 0.0001;
-        }
+#[derive(Clone)]
+pub struct RpcCfg {
+    pub url: String,
+    pub timeout: Duration,
+    pub commitment: CommitmentConfig,
+}
+
+/// Helper function for claiming the rate-limited mutex and constructing an RPC instance
+async fn lock_and_make_rpc(rlmtx: &RLMutex<RpcCfg>) -> RpcClient {
+    let RpcCfg {
+        url,
+        timeout,
+        commitment,
+    } = rlmtx.lock().await.clone();
+    RpcClient::new_with_timeout_and_commitment(url, timeout, commitment)
+}
 
-        let successes_pct = tx_successes as f64 / total * 100.0;
-        let send_failures_pct = tx_send_failures as f64 / total * 100.0;
-        let confirm_failures_pct = tx_confirm_failures as f64 / total * 100.0;
-
-        info!(
-            "Stats since start:
-Runtime: {}s
-TPS: {:.4} tx/s ({:.4} symbols/s)
-Total attestation attempts: {}
-Successful txs: {} ({:.2})%
-Sending failures: {} ({:.2})%
-Confirmation failures: {} ({:.2})%
-Average batch resend delay: {:.2}s",
-            stats_start_time.elapsed().as_secs(),
-            tps,
-            sym_ps,
-            total,
-            tx_successes,
-            successes_pct,
-            tx_send_failures,
-            send_failures_pct,
-            tx_confirm_failures,
-            confirm_failures_pct,
-            batch_wait_times.as_secs() as f64 / (tx_successes as f64 + 0.000001),
+/// A future that decides how a batch is sent.
+///
+/// In daemon mode, attestations of the batch are scheduled
+/// continuously using spawn(), which means that a next attestation
+/// does not have to wait until the previous completes. Subsequent
+/// attestations are made according to the attestation_conditions
+/// field on the batch.
+///
+/// With daemon_mode off, this future attempts only one blocking
+/// attestation of the batch and returns the result.
+async fn attestation_sched_job(
+    mut batch: BatchState<'_>,
+    batch_no: usize,
+    batch_count: usize,
+    n_retries: usize,
+    retry_interval: Duration,
+    daemon: bool,
+    rpc_cfg: Arc<RLMutex<RpcCfg>>,
+    p2w_addr: Pubkey,
+    config: Pyth2WormholeConfig,
+    payer: Keypair,
+) -> Result<(), ErrBoxSend> {
+    let mut retries_left = n_retries;
+    // Enforces the max batch job count
+    let sema = Arc::new(Semaphore::new(batch.conditions.max_batch_jobs));
```

Comment:
Having multiple in-flight instances of the same request (well, technically batch but you know what I mean) is possibly overkill - what's the rationale here? Did it solve a particular problem you ran into?

Reply:
The period from sending to confirmation will often take longer than the duration between subsequent crank turns of a batch. This changeset decouples scheduling a crank turn from waiting for its confirmations on Solana. This enables better attestation rates per-symbol.

Reply:
I will try rewording the comment for this function. But it's very hard to explain concisely